src/transformers/generation_contrastive_search.py

```diff
@@ -0,0 +1,196 @@
+# coding=utf-8
+# Copyright 2022 University of Cambridge, Tencent AI Lab, DeepMind and The University of Hong Kong Authors and The HuggingFace Inc. team. All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+This file contains the utils functions for the contrastive search, which will be called in `generation_utils`
+"""
+
+from typing import Tuple
+
+import torch
+from torch import nn
+
+from .utils import logging
+
+
+logger = logging.get_logger(__name__)
+
+
+def ranking_fast(
```

Comment:
If possible I'd be in favor of not adding a new file here. IMO `ranking_fast` can be changed to a logit processor and `ContrastiveDecodingOneStepFast` is could be fully moved into `generation_utils.py` right away. 

IMO this will be easier to maintain and understand for people that know already how `generate` works - wdyt @gante ?